extmod/modssl_mbedtls: Add cert verification callback.

felixdoerre opened the pull request:
the update of mbedtls has activated MBEDTLS_SSL_KEEP_PEER_CERTIFICATE. This results in the old getpeercert function not working anymore. On top, storing the certificate for the whole duration of the connection is wasting RAM.

This PR enables access to the certificate through a callback so it does not have to be stored if it is not needed, an can be inspected and analyzed in python while the connection is established, optionally aborting the connection.

fixes #5835

felixdoerre commented:
@dpgeorge I've rebased and updated this PR to match the conflicting changes from #13181.

dpgeorge commented:
This seems like a pretty nice feature to have.  But... it's not CPython compatible.  We are really trying to keep these modules implementing a subset of CPython behaviour, and this is quite a difference to CPython.

Does CPython provide any way to react to an incoming connection with a callback?

To break with CPython, the way forward would be to follow the ideas of the `deflate` module, which is a MicroPython specific module and then CPython compatible `gzip` and `zlib` modules are provided on top of / implemented using `deflate`.

So we could define a new module, eg `tls`, in which we can have whatever API we want.  And then implement the `ssl` module as a simple Python wrapper around `tls` which is CPython compatible.

Actually, in the case of TLS support, it's a pretty advanced feature that most users don't want to worry about.   Instead a user would see a higher-level interface like `asyncio.start_server()`, or `aiohttp`.  So it's likely that a user never uses the `ssl` module directly.  In that case we can replace it with a custom `tls` module, and then add features like the one proposed here.

Note that all the hard work we have done on the `ssl` module so far won't be lost.  We'd simply rename `ssl` to `tls`, and provide a very simple Python `ssl.py` wrapper around `tls`.  This is anyway what CPython does, it implements `ssl.py` in terms of the built-in `_ssl` library.

felixdoerre commented:
Yes, cpython does not such a feature: https://github.com/python/cpython/issues/75425
Maybe also, because it's lets prudent, because `getpeercert` is always available.

(Maybe as a quick sidenote why I want this feature: I want to have a TLS-Client (around mqtt) that verifies the server certificate, but instead of supplying a CA certificate, I want to pin the certificate and verify its hash in the callback. This usecase does not feel so niche to me.)

Thinking about the code as it exists a bit more critically in terms of compatibility, the current `ssl`-package is pretty far from being a subset of cpython (I've tested with cpython 3.10.2):

- `SSLContext.set_ciphers()` micropython: accepts a list of IANA-Ciphersuite names; cpython: expects a single string indicating an openssl cipher preference list
- `SSLContext.get_ciphers()` micropython: outputs a list of IANA-Ciphersuite names; cpython: outputs a dictionary with multiple fields per selected cipher
- `ssl.wrap_socket` many keywords don't match with the cpython variant, but this function is luckily deprecated
- `ssl.MBEDTLS_VERSION` does not exist in cpython
- `SSLContext.load_cert_chain` allows byte objects in micropython. This is also explicitly documented.
- `verify_mode` for clients defaults to `VERIFY_NONE` for backwards compatibility reasons, differing from cpython

I understand why all those differences are there, and don't find them unreasonable, but still they are differences from the cpython behavior.

The change from this PR is arguably verify compatible (in the sense, that nearly all cpython programs that worked previously will still work with this change), even more than many of the other noted differences, as adding an additional optional KV-parameter and property can be easily ignored.


Abstracting the cpython compatible subset-interface would mean:
- no get/set ciphers
- we should probably leave out ssl.wrap_socket as that is deprecated and later removed in cpython
- load_cert_chain only accepts file names
- `verify_mode` for client defaults to `CERT_REQUIRED`.

Replacing the current micropython `ssl`-module with this would break compatibility, is that intended?

Carglglz commented:
> Note that all the hard work we have done on the ssl module so far won't be lost. We'd simply rename ssl to tls, and provide a very simple Python ssl.py wrapper around tls. This is anyway what CPython does, it implements ssl.py in terms of the built-in _ssl library.

That was my idea when I started porting `SSLContext` (see the outdated https://github.com/micropython/micropython-lib/pull/697), so if there is no firmware size concerns, it may be the way to go.

felixdoerre commented:
So, I am looking at the tests, cpython, the differences between mbedtls and axtls and starting see more and more cpython/micropython differences. I also noted the linked package in micropython-lib before.

So I am starting to wonder how to balance these goals against each other:
- keep (full) compatibility with the current micropython interface
- increase/reach compatibility with a cpython-subset interface, by removing/hiding features that are not fully compatible.

Looking a bit at the state of compatibility, here is how I would "sort" features:
### cpython compatible
- `SSLContext.wrap_socket`, with `server_side` and `server_hostname`
- `SSLContext.load_verify_locations` with `cafile` and `cadata` 
- `SSLContext.load_cert_chain` with `certfile` and `keyfile` only accepting paths
- `SSLContext.verify_mode` taking `CERT_NONE`, `CERT_OPTIONAL`, `CERT_REQUIRED` defaulting to `CERT_REQUIRED` for client contexts, and `CERT_NONE` for server contexts.
- (Wrapped Socket) `setblocking`

### micropython specific
-  `SSLContext.wrap_socket` with `do_handshake_on_connect` (this behaves slightly different from cpython, so I would move it here.
- `SSLContext.load_cert_chain` with `certfile` and `keyfile` also accepting bytes
- `SSLContext.get_ciphers`/`SSLContext.set_ciphers`
- `ssl.MBEDTLS_VERSION` (and maybe add `AXTLS_VERSION`?)
- (and the verification callback)

### micropython legacy compat
- "micropython specific" + `ssl.wrap_socket`
- override the `SSLContext.verify_mode` default to `CERT_NONE`.

As next steps I would suggest the following:
- Move the implementation as-is to `tls` (or `_ssl` to mimic the current python behavior), removing `ssl.wrap_socket`
- Implement a thin wrapper to provide "micropython legacy compat", that does not hide internals, but adds `ssl.wrap_socket`, to keep compatibility for now.
- Document any non-cpython compatible behavior in `ssl` as deprecated (and/or remove it from the documentation)

After enough time has passed for the deprecation, i.e. code that uses `SSLContext.wrap_socket(do_handshake_on_connect=True`, or `ssl.wrap_socket` (like `asyncio.start_server()`) has had chance to migrate to `tls`/`_ssl`, remove the compat-`ssl`.
At this point we can decide if we want to embed the micropython-lib `ssl` module

What do you think of that plan? If it sounds good to you, I'd start implementing these changes.

felixdoerre commented:
I went ahead and implemented what I suggested. Compatibility with the current micropython-api is maintained through a python-implemented module. I've moved get/ciphers to the internal API, as that is pretty new and hopefully not already relied upon too much.

dpgeorge commented:
Thanks @felixdoerre for updating the PR, it looks reasonable although I need to review it in depth.

> Move the implementation as-is to `tls` (or `_ssl` to mimic the current python behavior),

I think we should stay away from `_ssl` as a name because it clashes with CPython's module of that name and obviously does not have the same methods/classes/API as the CPython version (eg it may confuse tests that try to import `_ssl`, which would pass on both CPython and MicroPython).

I think `tls` is a good name, and we can then promote this `tls` module as the way to do SSL/TLS in MicroPython.

As for the `ssl.py` wrapper, that should probably go in `micropython-lib`, like, eg, `gzip.py` wrapper around the `deflate` module.

Carglglz commented on extmod/modtls_mbedtls.c:
```diff
@@ -91,24 +92,6 @@ STATIC mp_obj_t ssl_socket_make_new(mp_obj_ssl_context_t *ssl_context, mp_obj_t
 /******************************************************************************/
 // Helper functions.
 
-STATIC mp_obj_t read_file(mp_obj_t self_in) {
```
@felixdoerre why would you delete this? I thought this was better (i.e. less code size/memory) than reading the files in `ssl.py` module? Have you compared both implementations?

felixdoerre replied:
I didn't until now, because for me the comparison was "have it implement in c" vs "don't have it in c", as the goal is to only have the tls module and eventually make `ssl.py` optional.

However I've just now tested it and it seems that the implementation in python is indeed smaller (and handler an error in `f.read` correctly, by closing the file:
```
   text    data     bss     dec     hex filename
 718248   75912    7056  801216   c39c0 build-standard/micropython (as implemented in this PR)
 718568   75944    7056  801568   c3b20 build-standard/micropython (with the file read logic moved back to C)
```

Carglglz replied:
Then I guess it makes sense to move it to `ssl.py`, thanks for testing it 👍🏼

dpgeorge replied:
Good to know that it's smaller in Python!

Carglglz commented:
@felixdoerre Could you add a test to `multi_net` to test the cert callback? (these tests are run in CI unix builds and it is where coverage result comes from)

felixdoerre commented:
Sure, I've added a "good case" test case for the cert callback there. I think that is enough and there is no need to also mirror the "fail case" there.

dpgeorge commented:
Link to corresponding `micropython-lib` changes: https://github.com/micropython/micropython-lib/pull/793

dpgeorge commented on tests/multi_net/sslcontext_certcallback.py:
```diff
@@ -0,0 +1,68 @@
+    import io
+    import os
+    import socket
+    import tls as ssl
```
Just do `import tls`.  No need to make it an alias.

dpgeorge commented on extmod/modssl_mbedtls.c:
```diff
@@ -270,13 +272,30 @@ STATIC mp_obj_t ssl_context_make_new(const mp_obj_type_t *type_in, size_t n_args
         // Load attribute.
         if (attr == MP_QSTR_verify_mode) {
             dest[0] = MP_OBJ_NEW_SMALL_INT(self->authmode);
+        } else if (attr == MP_QSTR_cert_callback) {
```
I suggest calling this attribute `verify_callback`.  It indicates that it's related to the existing `verify_mode` attribute, and matches the CPython WIP PR: https://github.com/python/cpython/pull/31391

dpgeorge commented:
If possible, can you please swap the order of commits around and do the renaming of `ssl` to `tls` first, then add the verification callback?  That seems more logical to me, and is easier to review.

felixdoerre committed:
extmod/modtls_mbedtls: Implement cert verification callback for mbedtls.

This is a useful alternative to .getpeercert() when the certificate is not
stored to reduce RAM usage.

Signed-off-by: Felix Dörre <felix@dogcraft.de>

felixdoerre committed:
extmod/modtls: Move the native ssl module to tls.

The current `ssl` module has quite a few differences to the CPython
implementation.  This change moves the MicroPython variant to a new `tls`
module and provides a wrapper module for `ssl` (in micropython-lib).

Users who only rely on implemented comparible behavior can continue to use
`ssl`, while users that rely on non-compatible behavior should switch to
`tls`.  Then we can make the facade in `ssl` more strictly adhere to
CPython.

Signed-off-by: Felix Dörre <felix@dogcraft.de>

felixdoerre committed:
extmod/modssl_axtls: Add SSLContext.load_cert_chain().

To match the mbedtls implementation.

Signed-off-by: Felix Dörre <felix@dogcraft.de>

felixdoerre commented:
Thanks for the review. I're incorporated all changes, and changed the order of commits.

dpgeorge commented on ports/unix/variants/nanbox/manifest.py:
```diff
@@ -0,0 +1 @@
+require("ssl")
```
This file is not needed.

dpgeorge commented on ports/windows/variants/standard/mpconfigvariant.props:
```diff
@@ -1,3 +1,6 @@
 <?xml version="1.0" encoding="utf-8"?>
 <Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003" ToolsVersion="14.0">
+  <PropertyGroup>
+    <FrozenManifest>$(PyVariantDir)manifest.py</FrozenManifest>
```
I suggest using `$(PyWinDir)\variants\manifest.py`, then the file `ports/windows/variants/standard/manifest.py` isn't needed.

dpgeorge commented on tests/multi_net/sslcontext_server_client_ciphers.py:
```diff
@@ -4,6 +4,7 @@
     import os
     import socket
     import ssl
+    import tls
```
Since this test is MicroPython-specific, may as well remove the `import ssl` and change all uses of `ssl` in this test to `tls` (should be easy).

dpgeorge commented on docs/library/ssl.rst:
```diff
@@ -13,7 +13,7 @@ facilities for network sockets, both client-side and server-side.
 ---------
 
-.. function:: ssl.wrap_socket(sock, server_side=False, keyfile=None, certfile=None, cert_reqs=CERT_NONE, cadata=None, server_hostname=None, do_handshake=True)
+.. function:: ssl.wrap_socket(sock, server_side=False, key=None, cert=None, cert_reqs=CERT_NONE, cadata=None, server_hostname=None, do_handshake=True)
```
Done in ac8e7f7b67ea13e7e9f816d3cfd4ddae4782db8d

dpgeorge commented on extmod/modtls_mbedtls.c:
```diff
@@ -250,13 +252,30 @@ STATIC mp_obj_t ssl_context_make_new(const mp_obj_type_t *type_in, size_t n_args
+    if (o->ssl_context->handler == mp_const_none) {
+        return 0;
+    }
+    return mp_obj_get_int(mp_call_function_2(o->ssl_context->handler, mp_obj_new_bytes(crt->raw.p, crt->raw.len), MP_OBJ_NEW_SMALL_INT(depth)));
```
I suggest using `mp_obj_new_bytearray_by_ref(ctr->raw.len, crt->raw.p)` (instead of `mp_obj_new_bytes`).  Using a bytearray by reference will save RAM by not creating a copy of the data.

The downside of using bytearray by reference is that the object (memory it points to) becomes invalid after the callback.  But, this is an advanced feature and we can document the behaviour.  We need to save memory wherever possible.

Also, is it worth passing through the `*flags` argument??  I'm not sure what this does though.

felixdoerre replied:
The flags argument is an in-out parameter that allows the inspection and modification of the "default" verification errors: https://github.com/Mbed-TLS/mbedtls/blob/32c28cebb41a25bb4e18d3a530b59c0e5929a324/include/mbedtls/x509.h#L91

felixdoerre replied:
So I thought its more hassle and confusion to pass that through, than it would actually be worth.

dpgeorge commented on extmod/modtls_mbedtls.c:
```diff
@@ -250,13 +252,30 @@ STATIC mp_obj_t ssl_context_make_new(const mp_obj_type_t *type_in, size_t n_args
+
+STATIC void ssl_context_set_verify_callback(mp_obj_ssl_context_t *self, mp_obj_t handler_in) {
+    if (handler_in != mp_const_none && !mp_obj_is_callable(handler_in)) {
+        mp_raise_ValueError(MP_ERROR_TEXT("invalid handler"));
```
I don't think it's necessary to validate the handler.  It's an advanced feature and users should pass in a callable.  We can document that and provide examples.

felixdoerre replied:
I'm not sure what happens when you `mp_call_function` something that is not callable. If that throws an error, sure there is no immediate need to verify.

dpgeorge replied:
Calls are always verified, so it's not necessary to validate it explicitly.

dpgeorge commented on extmod/modtls_mbedtls.c:
```diff
@@ -267,6 +286,9 @@ STATIC void ssl_context_attr(mp_obj_t self_in, qstr attr, mp_obj_t *dest) {
             mbedtls_ssl_conf_authmode(&self->conf, self->authmode);
+        } else if (attr == MP_QSTR_verify_callback) {
+            dest[0] = MP_OBJ_NULL;
+            ssl_context_set_verify_callback(self, dest[1]);
```
I don't think this needs to be a separate helper function (to set the callback).  It can simply be `self->handler = dest[1]`.

dpgeorge commented on extmod/modtls_mbedtls.c:
```diff
@@ -502,6 +524,7 @@ STATIC mp_obj_t ssl_socket_make_new(mp_obj_ssl_context_t *ssl_context, mp_obj_t
         mbedtls_ssl_free(&o->ssl);
         mp_raise_ValueError(MP_ERROR_TEXT("CERT_REQUIRED requires server_hostname"));
     }
+    mbedtls_ssl_set_verify(&o->ssl, &ssl_sock_cert_verify, o);
```
Why not use `mbedtls_ssl_conf_verify()` and call it in `ssl_context_make_new()`?

dpgeorge commented on tests/net_hosted/ssl_verify_callback.py:
```diff
@@ -0,0 +1,37 @@
+
+
+def verify_callback(cert, depth):
+    print(type(cert), len(cert) > 100, depth)
```
I suggest putting `"verify_callback"` at the start of this print, to distinguish lines in the output.

dpgeorge commented on tests/net_hosted/ssl_verify_callback.py:
```diff
@@ -0,0 +1,37 @@
+
+
+def verify_callback_fail(cert, depth):
+    print(type(cert), len(cert) > 100, depth)
```
I suggest putting "verify_callback_fail" at the start of this print, to distinguish lines in the output.

dpgeorge commented:
I'm not sure why the AppVeyor CI is failing.  Something to do with `micropython-lib` not being found.  Maybe I need to merge the `micropython-lib` changes first...

dpgeorge commented on extmod/modtls_mbedtls.c:
```diff
@@ -206,6 +189,14 @@ STATIC void ssl_check_async_handshake_failure(mp_obj_ssl_socket_t *sslsock, int
+    if (o->handler == mp_const_none) {
+        return 0;
+    }
+    return mp_obj_get_int(mp_call_function_2(o->handler, mp_obj_new_bytes(crt->raw.p, crt->raw.len), MP_OBJ_NEW_SMALL_INT(depth)));
```
What about changing the `mp_obj_new_bytes()` to `mp_obj_new_bytearray_by_ref()`?

dpgeorge commented on tests/net_hosted/ssl_verify_callback.py:
```diff
@@ -0,0 +1,37 @@
+
+
+def verify_callback(cert, depth):
+    print("verify_callback: ", type(cert), len(cert) > 100, depth)
```
This doesn't need a space after the colon (Python will add spaces between printed arguments).

Similarly with `verify_callback_fail` below.

dpgeorge commented on extmod/modtls_mbedtls.c:
```diff
@@ -206,6 +189,14 @@ STATIC void ssl_check_async_handshake_failure(mp_obj_ssl_socket_t *sslsock, int
+    if (o->handler == mp_const_none) {
+        return 0;
+    }
+    return mp_obj_get_int(mp_call_function_2(o->handler, mp_obj_new_bytearray_by_ref(crt->raw.len, crt->raw.p), MP_OBJ_NEW_SMALL_INT(depth)));
```
Actually, thinking about this some more, it's probable even better to use a `memoryview`, because then it can be made read-only.

This is how it's done for `bluetooth.BLE`, in the IRQ callback handler.

Eg:
```c
mp_obj_array_t cert;
mp_obj_memoryview_init(&cert, 'B', 0, crt->raw.len, crt->raw.len);
return mp_obj_get_int(mp_call_function_2(o->handler, MP_OBJ_NEW_FROM_PTR(&cert), MP_OBJ_NEW_SMALL_INT(depth)));
```

dpgeorge replied:
Thanks for updating with this suggestion.  Looks good now!

dpgeorge committed:
lib/micropython-lib: Update submodule to latest.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
github/workflows: Initialise micropython-lib submodule for windows CI.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
The Windows CI is now passing (it needed an explicit `git submodule update --init lib/micropython-lib`).

dpgeorge commented:
Now merged!

Thanks a lot for your efforts on this @felixdoerre .
